README.rst
====
istr
====


.. image:: https://img.shields.io/travis/saltstack/istr.svg
        :target: https://travis-ci.org/saltstack/istr

.. image:: https://ci.appveyor.com/api/projects/status/github/saltstack/istr?branch=master&svg=true
    :target: https://ci.appveyor.com/project/saltstack-public/istr/branch/master
    :alt: See Build Status on AppVeyor

.. image:: https://codecov.io/github/saltstack/istr/coverage.svg?branch=master
    :target: https://codecov.io/github/saltstack/istr?branch=master
    :alt: Code Coverage

.. image:: https://readthedocs.org/projects/istr/badge/?version=latest
        :target: https://istr.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/saltstack/istr/shield.svg
     :target: https://pyup.io/repos/github/saltstack/istr/
     :alt: Updates

.. image:: https://img.shields.io/pypi/v/istr.svg?style=flat
    :alt: PyPI Package latest release
    :target: https://pypi.python.org/pypi/istr

.. image:: https://img.shields.io/pypi/wheel/istr.svg?style=flat
    :alt: PyPI Wheel
    :target: https://pypi.python.org/pypi/istr

.. image:: https://img.shields.io/pypi/pyversions/istr.svg?style=flat
    :alt: Supported versions
    :target: https://pypi.python.org/pypi/istr

.. image:: https://img.shields.io/pypi/implementation/istr.svg?style=flat
    :alt: Supported implementations
    :target: https://pypi.python.org/pypi/istr


Case-insensitive string implementation based on:

* https://code.activestate.com/recipes/194371/
* https://stackoverflow.com/a/29247821/1258307


License
-------
* Free software: Apache Software License 2.0


Documentation
-------------

* https://istr.readthedocs.io.


Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage



HISTORY.rst
=======
History
=======

1.x.x (2020-xx-xx)
--------------------

1.x.x (2020-02-17)
--------------------

* Fixed upgrading of pip in tox.ini to use 'python -m pip install' so it can
  work on Windows.
* Added support for Python 3.6, 3.7, 3.8.

1.0.1 (2018-01-08)
--------------------

* Minor changes to README for proper description on PyPi


1.0.0 (2018-01-08)
------------------

* First release (Deleted from PyPI).


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for istr."""


tests/conftest.py
# -*- coding: utf-8 -*-
'''
    tests.conftest
    ~~~~~~~~~~~~~~
'''

# Import python libs
import os
import sys

CODE_ROOT = os.path.dirname(os.path.dirname(__file__))
if CODE_ROOT in sys.path:
    sys.path.remove(CODE_ROOT)
sys.path.insert(0, CODE_ROOT)


tests/test_match.py
# -*- coding: utf-8 -*-
'''
    tests.test_match
    ~~~~~~~~~~~~~~~~

    iStr unit tests
'''

# Import python libs
import os
import logging
from random import Random

# Import pytest libs
import pytest

# Import 3rd-party libs
from faker import Factory

# Import raas libs
from istr import istr

log = logging.getLogger(__name__)

SEED_NUMBER = 12345

LOCALES = (
    'en_US',
    # Now come the Unicode interesting locales
    'bg_BG',  # Bulgarian'
    'cs_cZ',  # Czech
    'de_DE',  # German
    'el_GR',  # Greek
    'fi_FI',  # Finish
    'lv_LV',  # Latvian
    'ru_RU',  # Russian
    'sv_SE',  # Swedish
    'tr_TR',  # Turkish
    'uk_UA',  # Ukrainian
    # The following would be interesting but apparently there's no
    # lower/upper case
    # 'fa_IR',  # Persian, Iran
    # 'hi_IN',  # Hindi
    # 'ja_JP',  # Japanese
    # 'ko_KR',  # Korean
    # 'ne_NP',  # Nepali
    # 'zh_CN',  # Chinese, China
    # 'zh_TW',  # Chinese, Taiwan
)

LOG_STRINGS = 'CI_RUN' not in os.environ


def munge_string_case(text, random_seed=None):
    random = Random(random_seed)
    munged = list(text)
    size = len(text)
    choices = list(range(size))
    while size >= 0:
        char_index = random.choice(choices)
        if char_index % 2 == 0:
            munged[char_index] = munged[char_index].upper()
        else:
            munged[char_index] = munged[char_index].lower()
        size -= 1
    munged = ''.join(munged)
    assert text != munged
    return munged


def _locale_to_fixture_id(value):
    return 'Locale({})'.format(value)


@pytest.fixture(params=LOCALES, ids=_locale_to_fixture_id)
def locale(request):
    return request.param


@pytest.fixture
def faker(locale):
    faker_instance = Factory.create(locale)
    # We want random, but expectable randomness
    faker_instance.random.seed(SEED_NUMBER)
    return faker_instance


def _sample_to_fixture_id(value):
    return 'Sample({})'.format(value)


@pytest.fixture(params=list(range(1, 11)), ids=_sample_to_fixture_id)
def sample(faker):
    original = faker.name()
    munged = munge_string_case(original, random_seed=SEED_NUMBER)
    return original, munged


@pytest.fixture(params=list(range(1, 11)), ids=_sample_to_fixture_id)
def two_samples(faker):
    original1 = faker.name()
    munged1 = munge_string_case(original1, random_seed=SEED_NUMBER)
    original2 = faker.name()
    munged2 = munge_string_case(original2, random_seed=SEED_NUMBER)
    assert original1 != original2
    return original1, munged1, original2, munged2


def test_repr(sample):
    text, _ = sample
    assert repr(istr(text)) == "istr({})".format(repr(text))


def test_eq_equal(sample):
    original, munged = sample
    if LOG_STRINGS:
        try:
            log.debug('original(%r) == munged(%r)',
                      original, munged)
        except UnicodeEncodeError:
            pass
    assert istr(original) == munged
    assert istr(original) == istr(munged)


def test_eq_different(two_samples):
    original1, _, _, munged2 = two_samples
    if LOG_STRINGS:
        try:
            log.debug('not original(%r) == munged(%r)',
                      original1, munged2)
        except UnicodeEncodeError:
            pass
    assert not istr(original1) == munged2
    assert not istr(original1) == istr(munged2)


def test_ne_equal(sample):
    original, munged = sample
    if LOG_STRINGS:
        try:
            log.debug('not original(%r) != munged(%r)',
                      original, munged)
        except UnicodeEncodeError:
            pass
    assert not istr(original) != munged
    assert not istr(original) != istr(munged)


def test_ne_different(two_samples):
    original1, _, _, munged2 = two_samples
    if LOG_STRINGS:
        try:
            log.debug('original(%r) != munged(%r)',
                      original1, munged2)
        except UnicodeEncodeError:
            pass
    assert istr(original1) != munged2
    assert istr(original1) != istr(munged2)


def test_lt(two_samples):
    original1, munged1, original2, munged2 = two_samples
    is_1_less = original1.casefold() < original2.casefold()
    if is_1_less:
        if LOG_STRINGS:
            try:
                log.debug('original(%r) < munged(%r)',
                          original1, munged2)
            except UnicodeEncodeError:
                pass
        assert istr(original1) < munged2
        assert istr(original1) < istr(munged2)
    else:
        if LOG_STRINGS:
            try:
                log.debug('original(%r) < munged(%r)',
                          original2, munged1)
            except UnicodeEncodeError:
                pass
        assert istr(original2) < munged1
        assert istr(original2) < istr(munged1)


def test_le_different(two_samples):
    original1, munged1, original2, munged2 = two_samples
    is_1_less = original1.casefold() < original2.casefold()
    if is_1_less:
        if LOG_STRINGS:
            try:
                log.debug('original(%r) <= munged(%r)',
                          original1, munged2)
            except UnicodeEncodeError:
                pass
        assert istr(original1) <= munged2
        assert istr(original1) <= istr(munged2)
    else:
        if LOG_STRINGS:
            try:
                log.debug('original(%r) <= munged(%r)',
                          original2, munged1)
            except UnicodeEncodeError:
                pass
        assert istr(original2) <= munged1
        assert istr(original2) <= istr(munged1)


def test_le_equal(sample):
    original, munged = sample
    if LOG_STRINGS:
        try:
            log.debug('original(%r) <= munged(%r)',
                      original, munged)
        except UnicodeEncodeError:
            pass
    assert istr(original) <= munged
    assert istr(original) <= istr(munged)


def test_gt(two_samples):
    original1, munged1, original2, munged2 = two_samples
    is_1_greater = original1.casefold() > original2.casefold()
    if is_1_greater:
        if LOG_STRINGS:
            try:
                log.debug('original(%r) > munged(%r)',
                          original1, munged2)
            except UnicodeEncodeError:
                pass
        assert istr(original1) > munged2
        assert istr(original1) > istr(munged2)
    else:
        if LOG_STRINGS:
            try:
                log.debug('original(%r) > munged(%r)',
                          original2, munged1)
            except UnicodeEncodeError:
                pass
        assert istr(original2) > munged1
        assert istr(original2) > istr(munged1)


def test_ge_different(two_samples):
    original1, munged1, original2, munged2 = two_samples
    is_1_greater = original1.casefold() > original2.casefold()
    if is_1_greater:
        if LOG_STRINGS:
            try:
                log.debug('original(%r) >= munged(%r)',
                          original1, munged2)
            except UnicodeEncodeError:
                pass
        assert istr(original1) >= munged2
        assert istr(original1) >= istr(munged2)
    else:
        if LOG_STRINGS:
            try:
                log.debug('original(%r) >= munged(%r)',
                          original2, munged1)
            except UnicodeEncodeError:
                pass
        assert istr(original2) >= munged1
        assert istr(original2) >= istr(munged1)


def test_ge_equal(sample):
    original, munged = sample
    if LOG_STRINGS:
        try:
            log.debug('original(%r) >= munged(%r)',
                      original, munged)
        except UnicodeEncodeError:
            pass
    assert istr(original) >= munged
    assert istr(original) >= istr(munged)


def test_hash(sample):
    original, munged = sample
    if LOG_STRINGS:
        try:
            log.debug('hash(original(%r)) == hash(munged(%r))',
                      original, munged)
        except UnicodeEncodeError:
            pass
    assert hash(original) != hash(munged)
    assert hash(istr(original)) == hash(istr(munged))


def test_contains(sample):
    original, munged = sample
    munged_part = munged[1:-1]
    if LOG_STRINGS:
        try:
            log.debug('munged_part(%r) in original(%r)',
                      munged_part, original)
        except UnicodeEncodeError:
            pass
    assert munged_part not in original
    assert munged_part in istr(original)
    assert istr(munged_part) in istr(original)


def test_count(sample):
    original, munged = sample
    munged_other = munged[1:-1]
    if LOG_STRINGS:
        try:
            log.debug('original(%r).count(munged_other(%r))',
                      original, munged_other)
        except UnicodeEncodeError:
            pass
    assert original.count(munged_other) == 0
    assert istr(original).count(munged_other) == 1
    assert istr(original).count(istr(munged_other)) == 1


def test_find(sample):
    original, munged = sample
    pos = 2
    munged_other = munged[pos:-1]
    if LOG_STRINGS:
        try:
            log.debug('original(%r).find(munged_other(%r))',
                      original, munged_other)
        except UnicodeEncodeError:
            pass
    assert original.find(munged_other) == -1
    assert istr(original).find(munged_other) == pos
    assert istr(original).find(istr(munged_other)) == pos


def test_index(sample):
    original, munged = sample
    pos = 2
    munged_other = munged[pos:-1]
    if LOG_STRINGS:
        try:
            log.debug('original(%r).index(munged_other(%r))',
                      original, munged_other)
        except UnicodeEncodeError:
            pass
    with pytest.raises(ValueError):
        assert original.index(munged_other)
    assert istr(original).index(munged_other) == pos
    assert istr(original).index(istr(munged_other)) == pos


def test_rfind(sample):
    original, munged = sample
    pos = 2
    munged_other = munged[pos:-1]
    if LOG_STRINGS:
        try:
            log.debug('original(%r).rfind(munged_other(%r))',
                      original, munged_other)
        except UnicodeEncodeError:
            pass
    assert original.rfind(munged_other) == -1
    assert istr(original).rfind(munged_other) == pos
    assert istr(original).rfind(istr(munged_other)) == pos


def test_rindex(sample):
    original, munged = sample
    pos = 2
    munged_other = munged[pos:-1]
    if LOG_STRINGS:
        try:
            log.debug('original(%r).rindex(munged_other(%r))',
                      original, munged_other)
        except UnicodeEncodeError:
            pass
    with pytest.raises(ValueError):
        assert original.rindex(munged_other)
    assert istr(original).rindex(munged_other) == pos
    assert istr(original).rindex(istr(munged_other)) == pos


def test_startswith(sample):
    original, munged = sample
    munged_other = munged[:-1]
    if LOG_STRINGS:
        try:
            log.debug('original(%r).startswith(munged_other(%r))',
                      original, munged_other)
        except UnicodeEncodeError:
            pass
    assert not original.startswith(munged_other)
    assert istr(original).startswith(munged_other)
    assert istr(original).startswith(istr(munged_other))


def test_endswith(sample):
    original, munged = sample
    munged_other = munged[1:]
    if LOG_STRINGS:
        try:
            log.debug('original(%r).endswith(munged_other(%r))',
                      original, munged_other)
        except UnicodeEncodeError:
            pass
    assert not original.endswith(munged_other)
    assert istr(original).endswith(munged_other)
    assert istr(original).endswith(istr(munged_other))


def test_casefold(sample):
    original, munged = sample
    if LOG_STRINGS:
        try:
            log.debug('original(%r).casefold() == munged(%r))',
                      original, munged)
        except UnicodeEncodeError:
            pass
    assert original.casefold() == munged.casefold()  # !!
    assert istr(original).casefold() == munged.casefold()
    assert istr(original).casefold() == istr(munged).casefold()


istr/__init__.py
# -*- coding: utf-8 -*-
'''
    istr.py
    ~~~~~~~

    Case insensitive string implementation.

    Based on https://code.activestate.com/recipes/194371/ and
    https://stackoverflow.com/a/29247821/1258307
'''

# Metadata
__author__ = 'Pedro Algarvio'
__email__ = 'pedro@algarvio.me'
__version__ = '1.0.2'
__version_info__ = tuple(
    int(part) for part in __version__.split('.') if part.isdigit()
)

# Import python libs
import unicodedata


class istr(str):  # pylint: disable=invalid-name
    '''
    Case insensitive strings class.
    Performs like str except comparisons are case insensitive.
    '''

    @property
    def normalized(self):
        if not hasattr(self, '__normalized'):
            text = super().__str__()
            try:
                # Can it be decoded using ascii
                text.encode('ascii')
                self.__normalized = text.casefold()
            except UnicodeEncodeError:
                # Text contains characters not in the ascii range
                self.__normalized = unicodedata.normalize('NFKC',
                                                          text.casefold())
        return self.__normalized

    def __repr__(self):
        return 'istr({})'.format(super().__repr__())

    def __eq__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return self.normalized == other.normalized

    def __lt__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return self.normalized < other.normalized

    def __le__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return self.normalized <= other.normalized

    def __gt__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return self.normalized > other.normalized

    def __ne__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return self.normalized != other.normalized

    def __ge__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return self.normalized >= other.normalized

    def __hash__(self):
        return hash(self.normalized)

    def __contains__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return other.normalized in self.normalized

    def count(self, other, *args):  # pylint: disable=arguments-differ
        if not isinstance(other, istr):
            other = istr(other)
        return str.count(self.normalized, other.normalized, *args)

    def endswith(self, other, *args):
        if not isinstance(other, istr):
            other = istr(other)
        return str.endswith(self.normalized, other.normalized, *args)

    def find(self, other, *args):  # pylint: disable=arguments-differ
        if not isinstance(other, istr):
            other = istr(other)
        return str.find(self.normalized, other.normalized, *args)

    def index(self, other, *args):  # pylint: disable=arguments-differ
        if not isinstance(other, istr):
            other = istr(other)
        return str.index(self.normalized, other.normalized, *args)

    def rfind(self, other, *args):
        if not isinstance(other, istr):
            other = istr(other)
        return str.rfind(self.normalized, other.normalized, *args)

    def rindex(self, other, *args):
        if not isinstance(other, istr):
            other = istr(other)
        return str.rindex(self.normalized, other.normalized, *args)

    def startswith(self, other, *args):
        if not isinstance(other, istr):
            other = istr(other)
        return str.startswith(self.normalized, other.normalized, *args)

    def casefold(self):
        return self.normalized.casefold()


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file."""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except ImportError:
    from urllib.request import urlopen


GITHUB_REPO = 'saltstack/istr'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key.

    Work around keys with incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning."""
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    """Load yaml config file at the given path."""
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    """Save yaml config file at the given path."""
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Put `encrypted_password` into the deploy section of .travis.yml."""
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    """Add a PyPI password to .travis.yml so that Travis can deploy to PyPI.

    Fetch the Travis public key for the repo, and encrypt the PyPI password
    with it before adding, so that only Travis can decrypt and use the PyPI
    password.
    """
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


src/istr.py
# -*- coding: utf-8 -*-
'''
    istr.py
    ~~~~~~~

    Case insensitive string implementation.

    Based on https://code.activestate.com/recipes/194371/ and
    https://stackoverflow.com/a/29247821/1258307
'''

# Metadata
__author__ = 'Pedro Algarvio'
__email__ = 'pedro@algarvio.me'
__version__ = '0.9.0'
__version_info__ = tuple(
    int(part) for part in __version__.split('.') if part.isdigit()
)

# Import python libs
import unicodedata


class istr(str):  # pylint: disable=invalid-name
    '''
    Case insensitive strings class.
    Performs like str except comparisons are case insensitive.
    '''

    @property
    def normalized(self):
        if not hasattr(self, '__normalized'):
            text = super().__str__()
            try:
                # Can it be decoded using ascii
                text.encode('ascii')
                self.__normalized = text.casefold()
            except UnicodeEncodeError:
                # Text contains characters not in the ascii range
                self.__normalized = unicodedata.normalize('NFKD',
                                                          text.casefold())
        return self.__normalized

    def __repr__(self):
        return 'istr({})'.format(super().__repr__())

    def __eq__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return self.normalized == other.normalized

    def __lt__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return self.normalized < other.normalized

    def __le__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return self.normalized <= other.normalized

    def __gt__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return self.normalized > other.normalized

    def __ne__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return self.normalized != other.normalized

    def __ge__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return self.normalized >= other.normalized

    def __hash__(self):
        return hash(self.normalized)

    def __contains__(self, other):
        if not isinstance(other, istr):
            other = istr(other)
        return other.normalized in self.normalized

    def count(self, other, *args):  # pylint: disable=arguments-differ
        if not isinstance(other, istr):
            other = istr(other)
        return str.count(self.normalized, other.normalized, *args)

    def endswith(self, other, *args):
        if not isinstance(other, istr):
            other = istr(other)
        return str.endswith(self.normalized, other.normalized, *args)

    def find(self, other, *args):  # pylint: disable=arguments-differ
        if not isinstance(other, istr):
            other = istr(other)
        return str.find(self.normalized, other.normalized, *args)

    def index(self, other, *args):  # pylint: disable=arguments-differ
        if not isinstance(other, istr):
            other = istr(other)
        return str.index(self.normalized, other.normalized, *args)

    def rfind(self, other, *args):
        if not isinstance(other, istr):
            other = istr(other)
        return str.rfind(self.normalized, other.normalized, *args)

    def rindex(self, other, *args):
        if not isinstance(other, istr):
            other = istr(other)
        return str.rindex(self.normalized, other.normalized, *args)

    def startswith(self, other, *args):
        if not isinstance(other, istr):
            other = istr(other)
        return str.startswith(self.normalized, other.normalized, *args)

    def casefold(self):
        return self.normalized.casefold()
